feat: add hidden documentation about SubscriptionCheckoutComponent (#115)
PrestaShopCorp/docs.cloud.prestashop.com: docs/src/5-prestashop-billing/100-hidden/1-checkout-subscription/README.md

---

[[toc]]

# Checkout Subscription

## Use case

Some of our users don't want to use the whole funnel, because it doesn't fit very well with the onboarding flow. We provide a checkout only component, which handle only the payment method, billing address and the subscription CTA.

This component can be imported in any layout: a page, a modal, a fullscreen modal...

![Checkout Subscription component](/assets/images/billing/checkout-subscription.png)

## Implementation

### Add the component in your page

First of all you need to add `@prestashopcorp/billing-cdc` to your project. 

```
# Import from CDN
<script src="https://unpkg.com/@prestashopcorp/billing-cdc/dist/bundle.js"></script>

# OR add it to your project dependencies
npm add @prestashopcorp/billing-cdc
```

`@prestashopcorp/billing-cdc` provides the component `SubscriptionCheckoutComponent`. Depending on your project setup, the way you will use this component will change.

#### Vue 3

You need to instantiate the component in the JS part, then use it in your Vue template.

::: warning
[Vite](https://vitejs.dev/) user can get this error `Uncaught TypeError: Expected string or element selector to be passed`. 

> In this cas an alias should be added in vite.config.js: `defineViteConfig({ resolve: { alias: { vue: "vue/dist/vue.esm-bundler.js" } } })`.

Reviewer: ```suggestion
In this case an alias should be added in vite.config.js: `defineViteConfig({ resolve: { alias: { vue: "vue/dist/vue.esm-bundler.js" } } })`.
```

```html
<body>
  <div id="ps-billing">
  <script>
    const subscription = new window.psBilling.SubscriptionCheckoutComponent({
      context: billingContext,
      onEventHook: eventHookHandler
    });
    customer.render('#ps-billing');
  </script>
</body>
```

### Build the context

We provide a PHP lib which create the billingContext for you, but if you want to display the checkout outside of a PrestaShop shop, you need to build it manually. Here is an example of the context. Please find more information about the context [in PrestaShop Billing CDC page](http://localhost:8080/5-prestashop-billing/6-references/3-billing-cdc/#context).

:::tip Note
The `contextType` set to shop means that you want to attach the subscription to a shop.
:::

```javascript
const billingContext = {
  contextType: 'shop',
  contextVersion: 2,
  billingEnv: 'production', // could be 'preprod' depending on your authentication configuration
  isSandbox: true, 
  accessToken: 't0K3N',     // the authentication token of your user, it should contain the scope "subscription.write"
  i18n: {
    isoCode: 'en',          // two-letter format
  },
  organization: {

>     uuid: 'authenticated_user_uUiD',

Reviewer: ```suggestion
    uuid: 'authenticated_user_uuid',
```

    email: 'john.doe@prestashop.com'
  },
  // This shop part is not required it the context you use is not "shop"
  shop: {
    uuid: 'shop_uuid',      // required if the shop node exists
    domain: '',             // the domain of the shop (optional)
  },
  product: {
    id: 'your_product_id',
    displayName: 'Product', // Used in the top of the summary card (optional)
    logoSrc: '',            // Used in the top of the summary card (optional)
    tosUrl: '',             // Link to your toc, displayed behind the summary card
    privacyUrl: '',         // Link to your privacy, displayed behind the summary card
  },
  offerSelection: {
    offerPricingId: 'the _plan_selected'
  },
}
```

> Please ask us to know the value to provide as `product.id` and `offerSelection.offerPricingId`.

Reviewer: ```suggestion
If you do not know what value to provide for the `product.id` and the`offerSelection.offerPricingId`, don't hesitate to ask us.
```

### React to the event hook

The event hook allows you to update your UI, when an important event occurs, like the creation of a subscription.

:::warning 

> Please do not use the event hook to register the fact a customer has subscribe to your product. It's not safe. Prefer the [webhook mechanism](../../6-references/1-webhook/).

Reviewer: ```suggestion
Please do not use the event hook to register the fact a customer has subscribed to your product. It's not safe. Prefer the [webhook mechanism](../../6-references/1-webhook/).
```

:::

> #### Vue 3

Reviewer: issue: pour les deux use-cases, je mettrais un rappel de comment ajouter les hooks sur l'initialization

Reviewer: Yep nickel
